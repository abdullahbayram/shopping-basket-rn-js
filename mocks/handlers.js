// eslint-disable-next-line import/no-extraneous-dependencies
import { http, HttpResponse } from 'msw';
import { products } from './fixtures';
import { checkBasket, checkCardNumber } from './handlers.util';

export const handlers = [
  http.get('http://localhost:9001/products', async () => {
    // await delay();
    return HttpResponse.json(products);
  }),

  http.post('http://localhost:9001/checkout', async ({ request }) => {
    // await delay(); // Uncomment this line to simulate network delay(Be careful, it will slow down your tests and may fail them)
    const body = await request.json();
    const { cardNumber, basket } = body;

    // Validate basket and card number
    const errors = [];
    const basketError = checkBasket(basket);
    const cardError = checkCardNumber(cardNumber);

    if (basketError) errors.push(basketError);
    if (cardError) errors.push(cardError);

    if (errors.length > 0) {
      return HttpResponse.json({ errors }, { status: 400 });
    }

    return HttpResponse.json({ message: 'The transaction was completed successfully.' });
  }),

  http.post('http://localhost:9001/promocode', async ({ request }) => {
    // await delay();
    const { promoCode } = await request.json();
    const match = promoCode?.match(/^A([\d]{1,2})$/);

    if (match) {
      const amount = parseInt(match[1], 10);
      return HttpResponse.json({ discountType: 'percent', amount });
    }

    return HttpResponse.json(
      { errors: [{ field: 'promoCode', message: 'The promo code you supplied is invalid.' }] },
      { status: 400 },
    );
  }),
];

export const sampleBasket = [
  {
    id: 1,
    title: 'Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops',
    price: 109.95,
    description:
      'Your perfect pack for everyday use and walks in the forest. Stash your laptop (up to 15 inches) in the padded sleeve, your everyday',
    category: "men's clothing",
    image: 'https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg',
    rating: {
      rate: 3.9,
      count: 120,
    },
    quantity: 3,
  },
  {
    id: 2,
    title: 'Mens Casual Premium Slim Fit T-Shirts ',
    price: 22.3,
    description:
      'Slim-fitting style, contrast raglan long sleeve, three-button henley placket, light weight & soft fabric for breathable and comfortable wearing. And Solid stitched shirts with round neck made for durability and a great fit for casual fashion wear and diehard baseball fans. The Henley style round neckline includes a three-button placket.',
    category: "men's clothing",
    image: 'https://fakestoreapi.com/img/71-3HjGNDUL._AC_SY879._SX._UX._SY._UY_.jpg',
    rating: {
      rate: 4.1,
      count: 259,
    },
    quantity: 4,
  },
  {
    id: 3,
    title: 'Mens Cotton Jacket',
    price: 55.99,
    description:
      'great outerwear jackets for Spring/Autumn/Winter, suitable for many occasions, such as working, hiking, camping, mountain/rock climbing, cycling, traveling or other outdoors. Good gift choice for you or your family member. A warm hearted love to Father, husband or son in this thanksgiving or Christmas Day.',
    category: "men's clothing",
    image: 'https://fakestoreapi.com/img/71li-ujtlUL._AC_UX679_.jpg',
    rating: {
      rate: 4.7,
      count: 500,
    },
    quantity: 2,
  },
  {
    id: 4,
    title: 'Mens Casual Slim Fit',
    price: 15.99,
    description:
      'The color could be slightly different between on the screen and in practice. / Please note that body builds vary by person, therefore, detailed size information should be reviewed below on the product description.',
    category: "men's clothing",
    image: 'https://fakestoreapi.com/img/71YXzeOuslL._AC_UY879_.jpg',
    rating: {
      rate: 2.1,
      count: 430,
    },
    quantity: 2,
  },
  {
    id: 5,
    title: "John Hardy Women's Legends Naga Gold & Silver Dragon Station Chain Bracelet",
    price: 695,
    description:
      "From our Legends Collection, the Naga was inspired by the mythical water dragon that protects the ocean's pearl. Wear facing inward to be bestowed with love and abundance, or outward for protection.",
    category: 'jewelery',
    image: 'https://fakestoreapi.com/img/71pWzhdJNwL._AC_UL640_QL65_ML3_.jpg',
    rating: {
      rate: 4.6,
      count: 400,
    },
    quantity: 3,
  },
];
