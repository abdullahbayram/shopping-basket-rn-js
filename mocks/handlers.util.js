export function createError(field, message) {
  return { field, message };
}

export function delay(minMs = 100, maxMs = 500) {
  // if (process.env.NODE_ENV === 'test') return Promise.resolve(); // Uncomment this line to speed up tests after implementing environment variables
  const timeout = Math.floor(Math.random() * (maxMs - minMs + 1) + minMs);
  return new Promise((resolve) => {
    setTimeout(resolve, timeout);
  });
}

export function checkCardNumber(cardNumber) {
  if (typeof cardNumber !== 'string') {
    return createError('cardNumber', 'You have not provided a card number.');
  }
  if (cardNumber.length !== 16) {
    return createError('cardNumber', 'You have not provided a valid card number.');
  }
  if (/^5249045959484101$/.test(cardNumber)) {
    return createError('cardNumber', 'Card can not be processed.');
  }
  return null;
}

export function checkBasket(basket) {
  if (!basket) {
    return createError('basket', 'You have not sent a basket.');
  }
  if (!Array.isArray(basket)) {
    return createError('basket', 'Basket has to be an array.');
  }
  if (basket.length === 0) {
    return createError('basket', 'Your basket is empty.');
  }

  // eslint-disable-next-line no-restricted-syntax
  for (const item of basket) {
    if (!item.id || !item.quantity) {
      return createError(
        'basket',
        'One or more items in your basket is invalid. Each item must have an id and quantity.',
      );
    }
  }
  return null;
}
