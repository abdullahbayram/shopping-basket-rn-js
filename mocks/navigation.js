const mockNavigation = {
  navigate: jest.fn(),
  goBack: jest.fn(),
  setParams: jest.fn(),
  navigation: {},
};

export default mockNavigation;
