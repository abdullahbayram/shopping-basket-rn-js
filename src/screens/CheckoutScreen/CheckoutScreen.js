import React from 'react';
import { Alert, FlatList, StyleSheet, View } from 'react-native';
import { useDispatch, useSelector } from 'react-redux';
import Button from '../../components/atoms/Button/Button';
import Text from '../../components/atoms/Text/Text';
import Screen from '../../components/templetes/Screen';
import Input from '../../components/atoms/Input/Input';
import ProductCard from '../../components/molecules/ProductCard/ProductCard';
import { removeItemFromBasket, clearBasket, updateItemQuantity } from '../../redux/slices/basketSlice';
import validateCreditCard from '../../utils/validateCreditCard'; //
import { usePlaceOrderMutation } from '../../redux/api/apiSlice';
import { selectBasketItems, selectTotalItemCount, selectTotalPrice } from '../../redux/selectors/basketSelector';

const CREDIT_CARD_CHECK = 'credit-card-check';
const CREDIT_CARD = 'credit-card';

const CheckoutScreen = () => {
  const items = useSelector(selectBasketItems);
  const totalCount = useSelector(selectTotalItemCount);
  const total = useSelector(selectTotalPrice);
  const dispatch = useDispatch();
  const [creditCardNumber, setCreditCardNumber] = React.useState('');
  const [isCreditCardValid, setIsCreditCardValid] = React.useState(false);

  const [placeOrder, { isLoading, isSuccess, isError, error }] = usePlaceOrderMutation();
  const onRemoveItem = (item) => {
    dispatch(removeItemFromBasket(item.sku));
  };

  const onPlaceOrder = async () => {
    if (!isCreditCardValid) {
      Alert.alert('Invalid Credit Card', 'Please enter a valid credit card number.');
    } else {
      try {
        const response = await placeOrder({
          basket: items,
          cardNumber: creditCardNumber,
        }).unwrap();
        if (isSuccess) {
          Alert.alert('Order Success', response.msg);
          dispatch(clearBasket());
          setCreditCardNumber('');
        }
        if (isError) {
          console.log(error);
          Alert.alert('Error', error);
        }
      } catch (err) {
        Alert.alert('Order Failed', err?.data || 'Failed to place order');
      }
    }
  };

  const onQuantityChange = (item, newQuantity) => {
    dispatch(updateItemQuantity({ sku: item.sku, quantity: newQuantity })); // Dispatch action to update quantity
  };

  React.useEffect(() => {
    setIsCreditCardValid(validateCreditCard(creditCardNumber));
  }, [creditCardNumber]);

  return (
    <Screen>
      <Text variant="titleMedium">Items in the basket: {totalCount}</Text>

      <FlatList
        data={items}
        renderItem={({ item }) => (
          <ProductCard
            key={item?.sku}
            title={item.name}
            subtitle={`${item.description} (x${item.quantity})`}
            buttonTitle="Remove Item"
            price={`${(item.price * item.quantity).toFixed(2)}`}
            onButtonPress={() => onRemoveItem(item)}
            keyExtractor={() => item.sku}
            quantity={item.quantity}
            onQuantityChange={(newQuantity) => onQuantityChange(item, newQuantity)}
          />
        )}
        keyExtractor={(item) => item.sku}
      />

      <View style={styles.totalContainer}>
        <Text variant="titleMedium">Total: ${total.toFixed(2)}</Text>
      </View>

      <Input
        value={creditCardNumber}
        onChangeText={setCreditCardNumber}
        icon={isCreditCardValid ? CREDIT_CARD_CHECK : CREDIT_CARD}
        label="Credit Card"
        placeholder="Enter your credit card number"
        maxLength={16}
        keyboardType="numeric"
      />

      <Button
        icon="cart-arrow-down"
        mode="contained"
        onPress={onPlaceOrder}
        disabled={items.length === 0 || !isCreditCardValid || isLoading}
      >
        ORDER
      </Button>
    </Screen>
  );
};

export default CheckoutScreen;

const styles = StyleSheet.create({
  totalContainer: {
    marginVertical: 16,
  },
});
