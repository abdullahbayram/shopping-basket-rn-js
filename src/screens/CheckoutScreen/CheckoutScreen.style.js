import { StyleSheet } from 'react-native';
import { spacing } from '../../constants/theme';

export default StyleSheet.create({
  totalContainer: {
    marginBottom: spacing.special,
    paddingLeft: spacing.xxs,
  },
  topContainer: {
    height: 165,
    justifyContent: 'space-between',
    marginVertical: spacing.sm,
  },
  orderButtonContainer: {
    flex: 4,
  },
  promoContainer: { flex: 6 },
});
