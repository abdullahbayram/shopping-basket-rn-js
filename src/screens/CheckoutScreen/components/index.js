export { default as TotalSummary } from './TotalSummary';
export { default as PromoCodeInput } from './PromoCodeInput';
