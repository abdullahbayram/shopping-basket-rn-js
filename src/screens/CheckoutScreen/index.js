import React from 'react';
import { View } from 'react-native';
import { TextInput, useTheme } from 'react-native-paper';
import { useDispatch, useSelector } from 'react-redux';
import { useForm, Controller } from 'react-hook-form';
import { Button, Text } from '../../components/atoms';
import { Input, ActivityOverlay } from '../../components/molecules';
import { CheckoutList } from '../../components/organisms';
import { BaseScreen } from '../../components/templetes';
import { removeItemFromBasket, updateItemQuantity, setDiscount } from '../../redux/slices/basketSlice';
import { useValidatePromoCodeMutation } from '../../redux/api/apiSlice';
import { selectBasketItems, selectTotalItemCount, selectTotalPrice } from '../../redux/selectors/basketSelector';
import validateBasket from '../../validate/validateBasket';
import showToast from '../../utils/showToast';
import messages from '../../constants/toastMessages';
import strings from '../../constants/strings';
import styles from './CheckoutScreen.style';

const CheckoutScreen = ({ navigation }) => {
  const { colors } = useTheme();
  const dispatch = useDispatch();

  const {
    control,
    handleSubmit,
    register,
    formState: { errors },
  } = useForm({
    defaultValues: {
      promoCode: '',
    },
  });

  const { basketItems, totalCount, total } = useSelector((state) => ({
    basketItems: selectBasketItems(state),
    totalCount: selectTotalItemCount(state),
    total: selectTotalPrice(state),
  }));

  const isBasketEmpty = basketItems.length === 0;

  const [validatePromoCode, { isLoading }] = useValidatePromoCodeMutation();

  const onRemoveItem = (item) => {
    dispatch(removeItemFromBasket(item.id));
  };

  const onOrderPress = () => {
    if (!validateBasket(basketItems)) {
      showToast(messages.basket.empty);
      return;
    }
    navigation.navigate('Payment');
  };

  const onApplyPromoCode = async (data) => {
    try {
      const { amount } = await validatePromoCode(data.promoCode).unwrap();
      if (amount) {
        dispatch(setDiscount(amount));
        showToast(messages.promo.success);
      } else {
        showToast(messages.promo.invalid);
      }
    } catch (err) {
      const errorMsg = err?.msg || messages.promoError.msg;
      showToast({ ...messages.promoError, msg: errorMsg });
    }
  };

  const onQuantityChange = (item, newQuantity) => {
    dispatch(updateItemQuantity({ id: item.id, quantity: newQuantity }));
  };

  return (
    <BaseScreen>
      <ActivityOverlay isVisible={isLoading} color={colors.secondary} />
      <View style={styles.totalContainer}>
        <Text variant="titleMedium">
          {strings.checkout.total} ${Number.isNaN(total) ? '0.00' : total.toFixed(2)}
        </Text>
      </View>

      <View style={styles.topContainer}>
        <View style={styles.orderButtonContainer}>
          <Button icon="cart-arrow-down" mode="contained" onPress={onOrderPress} disabled={isBasketEmpty}>
            {strings.checkout.order} ({totalCount} items)
          </Button>
        </View>

        <View style={styles.promoContainer}>
          <Controller
            control={control}
            render={({ field: { onChange, onBlur, value } }) => (
              <Input
                {...register('promoCode', {
                  required: 'Promo code can not be empty',
                  pattern: { value: /^A([\d]{1,2})$/, message: 'This promo code is not valid' },
                })}
                value={value}
                onChangeText={onChange}
                onBlur={onBlur}
                label="Promo Code"
                placeholder={strings.checkout.promoCodePlaceholder}
                keyboardType="default"
                right={<TextInput.Icon icon="percent" color={errors.promoCode ? colors.error : colors.primary} />}
                errorObject={errors.promoCode}
              />
            )}
            name="promoCode"
          />

          <Button icon="sack-percent" onPress={handleSubmit(onApplyPromoCode)} mode="contained">
            {strings.checkout.applyPromo}
          </Button>
        </View>
      </View>
      <CheckoutList onQuantityChange={onQuantityChange} basketItems={basketItems} onRemoveItem={onRemoveItem} />
    </BaseScreen>
  );
};

export default CheckoutScreen;
