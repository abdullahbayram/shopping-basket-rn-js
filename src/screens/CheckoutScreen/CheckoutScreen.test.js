import React from 'react';
import { fireEvent, screen } from '@testing-library/react-native';
import CheckoutScreen from './CheckoutScreen';
import renderInProvider from '../../../__tests__/utils/renderInProvider';
import { sampleBasket } from '../../../__tests__/mocks/handlers';

const initialState = { basket: { items: sampleBasket } };

describe('CheckoutScreen', () => {
  it('should render CheckoutScreen correctly', () => {
    renderInProvider(<CheckoutScreen />, { initialState });
    expect(screen.getByText('Items in the basket: 5')).toBeTruthy();
    expect(screen.getAllByText('Credit Card').length).toBe(2); // active and inactive text
    expect(screen.getByText('Product Two')).toBeTruthy();
    expect(screen.getByText('Product One description (x3)')).toBeTruthy();
    // expect(screen.getByText('Remove Item')).toBeTruthy();
    expect(screen.getByText('ORDER')).toBeTruthy();
  });
  it('should match the snapshot', () => {
    renderInProvider(<CheckoutScreen />);
    expect(screen.toJSON()).toMatchSnapshot();
  });
  it('should update credit card input value', () => {
    renderInProvider(<CheckoutScreen />, { initialState });

    const creditCardInput = screen.getByTestId('text-input-flat');
    fireEvent.changeText(creditCardInput, '1234 5678 9012 3456');

    expect(creditCardInput.props.value).toBe('1234 5678 9012 3456');
  });

  // remove item
});
