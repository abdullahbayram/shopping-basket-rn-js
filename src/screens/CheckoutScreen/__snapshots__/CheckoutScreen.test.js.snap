// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckoutScreen should match the snapshot 1`] = `
<View
  style={
    {
      "backgroundColor": "#fff",
      "flex": 1,
      "paddingLeft": 16,
      "paddingRight": 9,
      "paddingTop": 10,
    }
  }
>
  <View
    style={
      {
        "marginBottom": 7,
        "paddingLeft": 2,
      }
    }
  >
    <Text
      style={
        [
          {
            "textAlign": "left",
          },
          {
            "color": "rgba(28, 27, 31, 1)",
            "writingDirection": "ltr",
          },
          [
            {
              "fontFamily": "System",
              "fontSize": 16,
              "fontWeight": "500",
              "letterSpacing": 0.15,
              "lineHeight": 24,
            },
            [
              {
                "color": "rgba(125, 82, 96, 1)",
              },
              undefined,
            ],
          ],
        ]
      }
    >
      Total:
       $
      0.00
    </Text>
  </View>
  <View
    style={
      {
        "height": 165,
        "justifyContent": "space-between",
        "marginVertical": 10,
      }
    }
  >
    <View
      style={
        {
          "flex": 4,
        }
      }
    >
      <View
        collapsable={false}
        style={
          {
            "backgroundColor": "rgba(28, 27, 31, 0.12)",
            "borderRadius": 20,
            "shadowColor": "#000",
            "shadowOffset": {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 0,
            "shadowRadius": 0,
          }
        }
        testID="button-container-outer-layer"
      >
        <View
          collapsable={false}
          style={
            {
              "backgroundColor": "rgba(28, 27, 31, 0.12)",
              "borderColor": "transparent",
              "borderRadius": 20,
              "borderStyle": "solid",
              "borderWidth": 0,
              "flex": undefined,
              "minWidth": 64,
              "shadowColor": "#000",
              "shadowOffset": {
                "height": 0,
                "width": 0,
              },
              "shadowOpacity": 0,
              "shadowRadius": 0,
            }
          }
          testID="button-container"
        >
          <View
            accessibilityRole="button"
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": true,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              [
                {
                  "overflow": "hidden",
                },
                {
                  "borderRadius": 20,
                },
              ]
            }
            testID="button"
          >
            <View
              style={
                [
                  {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "center",
                  },
                  undefined,
                ]
              }
            >
              <View
                style={
                  [
                    {
                      "marginLeft": 12,
                      "marginRight": -4,
                    },
                    {
                      "marginLeft": 16,
                      "marginRight": -16,
                    },
                    false,
                  ]
                }
                testID="button-icon-container"
              >
                <Text
                  accessibilityElementsHidden={true}
                  allowFontScaling={false}
                  importantForAccessibility="no-hide-descendants"
                  pointerEvents="none"
                  selectable={false}
                  style={
                    [
                      {
                        "color": "rgba(28, 27, 31, 0.38)",
                        "fontSize": 18,
                      },
                      [
                        {
                          "lineHeight": 18,
                          "transform": [
                            {
                              "scaleX": 1,
                            },
                          ],
                        },
                        {
                          "backgroundColor": "transparent",
                        },
                      ],
                      {
                        "fontFamily": "material-community",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      {},
                    ]
                  }
                >
                  ó°µ¦
                </Text>
              </View>
              <Text
                numberOfLines={1}
                selectable={false}
                style={
                  [
                    {
                      "textAlign": "left",
                    },
                    {
                      "color": "rgba(28, 27, 31, 1)",
                      "writingDirection": "ltr",
                    },
                    [
                      {
                        "fontFamily": "System",
                        "fontSize": 14,
                        "fontWeight": "500",
                        "letterSpacing": 0.1,
                        "lineHeight": 20,
                      },
                      [
                        {
                          "marginHorizontal": 16,
                          "marginVertical": 9,
                          "textAlign": "center",
                        },
                        false,
                        {
                          "marginHorizontal": 24,
                          "marginVertical": 10,
                        },
                        undefined,
                        false,
                        {
                          "color": "rgba(28, 27, 31, 0.38)",
                          "fontFamily": "System",
                          "fontSize": 14,
                          "fontWeight": "500",
                          "letterSpacing": 0.1,
                          "lineHeight": 20,
                        },
                        undefined,
                      ],
                    ],
                  ]
                }
                testID="button-text"
              >
                ORDER
                 (
                0
                 items)
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
    <View
      style={
        {
          "flex": 6,
        }
      }
    >
      <View
        style={
          {
            "marginBottom": 5,
          }
        }
      >
        <View
          style={
            [
              {
                "backgroundColor": "rgba(231, 224, 236, 1)",
                "borderTopLeftRadius": 4,
                "borderTopRightRadius": 4,
              },
              {},
            ]
          }
        >
          <View
            collapsable={false}
            style={
              {
                "backgroundColor": "rgba(73, 69, 79, 1)",
                "bottom": 0,
                "height": 1,
                "left": 0,
                "position": "absolute",
                "right": 0,
                "transform": [
                  {
                    "scaleY": 0.5,
                  },
                ],
                "zIndex": 1,
              }
            }
            testID="text-input-underline"
          />
          <View
            onLayout={[Function]}
            style={
              [
                {
                  "flexGrow": 1,
                  "paddingBottom": 0,
                  "paddingTop": 0,
                },
                {
                  "minHeight": 56,
                },
              ]
            }
          >
            <View
              pointerEvents="none"
              style={
                [
                  {
                    "bottom": 0,
                    "left": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                  },
                  {
                    "overflow": "hidden",
                  },
                ]
              }
            >
              <View
                collapsable={false}
                pointerEvents="none"
                style={
                  {
                    "bottom": 0,
                    "left": 0,
                    "opacity": 1,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                    "transform": [
                      {
                        "translateX": 0,
                      },
                    ],
                    "width": 750,
                    "zIndex": 3,
                  }
                }
              >
                <View
                  style={
                    {
                      "width": 73,
                    }
                  }
                >
                  <Text
                    collapsable={false}
                    maxFontSizeMultiplier={1.5}
                    numberOfLines={1}
                    onLayout={[Function]}
                    onTextLayout={[Function]}
                    style={
                      {
                        "color": "rgba(103, 80, 164, 1)",
                        "fontFamily": "System",
                        "fontSize": 16,
                        "fontWeight": undefined,
                        "left": 0,
                        "letterSpacing": 0.15,
                        "lineHeight": undefined,
                        "maxWidth": 73,
                        "opacity": 0,
                        "paddingLeft": 16,
                        "paddingRight": 56,
                        "position": "absolute",
                        "textAlign": "left",
                        "top": 30,
                        "transform": [
                          {
                            "translateX": 0,
                          },
                          {
                            "translateY": 0,
                          },
                          {
                            "scale": 1,
                          },
                        ],
                        "writingDirection": "ltr",
                      }
                    }
                    testID="text-input-flat-label-active"
                  >
                    Promo Code
                  </Text>
                  <Text
                    collapsable={false}
                    maxFontSizeMultiplier={1.5}
                    numberOfLines={1}
                    style={
                      {
                        "color": "rgba(73, 69, 79, 1)",
                        "fontFamily": "System",
                        "fontSize": 16,
                        "fontWeight": undefined,
                        "left": 0,
                        "letterSpacing": 0.15,
                        "lineHeight": undefined,
                        "maxWidth": 73,
                        "opacity": 0,
                        "paddingLeft": 16,
                        "paddingRight": 56,
                        "position": "absolute",
                        "textAlign": "left",
                        "top": 30,
                        "transform": [
                          {
                            "translateX": 0,
                          },
                          {
                            "translateY": 0,
                          },
                          {
                            "scale": 1,
                          },
                        ],
                        "writingDirection": "ltr",
                      }
                    }
                    testID="text-input-flat-label-inactive"
                  >
                    Promo Code
                  </Text>
                </View>
              </View>
            </View>
            <TextInput
              cursorColor="rgba(103, 80, 164, 1)"
              editable={true}
              maxFontSizeMultiplier={1.5}
              multiline={false}
              onBlur={[Function]}
              onChangeText={[Function]}
              onFocus={[Function]}
              placeholder=" "
              placeholderTextColor="rgba(73, 69, 79, 1)"
              selectionColor="rgba(103, 80, 164, 1)"
              style={
                [
                  {
                    "flexGrow": 1,
                    "margin": 0,
                  },
                  {},
                  {
                    "paddingBottom": 4,
                    "paddingTop": 24,
                  },
                  {
                    "color": "rgba(28, 27, 31, 1)",
                    "fontFamily": "System",
                    "fontSize": 16,
                    "fontWeight": undefined,
                    "letterSpacing": 0.15,
                    "lineHeight": undefined,
                    "minWidth": 65,
                    "paddingLeft": 16,
                    "paddingRight": 56,
                    "textAlign": "left",
                    "textAlignVertical": "center",
                  },
                  false,
                  {
                    "marginRight": 40,
                    "paddingRight": 16,
                  },
                  undefined,
                ]
              }
              testID="text-input-flat"
              underlineColorAndroid="transparent"
              value=""
            />
          </View>
          <View
            style={
              [
                {
                  "alignItems": "center",
                  "height": 24,
                  "justifyContent": "center",
                  "position": "absolute",
                  "width": 24,
                },
                {
                  "right": 16,
                  "top": 16,
                },
              ]
            }
          >
            <View
              collapsable={false}
              style={
                {
                  "backgroundColor": "transparent",
                  "borderRadius": 20,
                  "height": 40,
                  "margin": 0,
                  "shadowColor": "#000",
                  "shadowOffset": {
                    "height": 0,
                    "width": 0,
                  },
                  "shadowOpacity": 0,
                  "shadowRadius": 0,
                  "width": 40,
                }
              }
              testID="right-icon-adornment-container-outer-layer"
            >
              <View
                collapsable={false}
                style={
                  {
                    "backgroundColor": "transparent",
                    "borderColor": "rgba(121, 116, 126, 1)",
                    "borderRadius": 20,
                    "borderWidth": 0,
                    "elevation": 0,
                    "flex": 1,
                    "overflow": "hidden",
                    "shadowColor": "#000",
                    "shadowOffset": {
                      "height": 0,
                      "width": 0,
                    },
                    "shadowOpacity": 0,
                    "shadowRadius": 0,
                  }
                }
                testID="right-icon-adornment-container"
              >
                <View
                  accessibilityComponentType="button"
                  accessibilityRole="button"
                  accessibilityState={
                    {
                      "busy": undefined,
                      "checked": undefined,
                      "disabled": false,
                      "expanded": undefined,
                      "selected": undefined,
                    }
                  }
                  accessibilityTraits="button"
                  accessibilityValue={
                    {
                      "max": undefined,
                      "min": undefined,
                      "now": undefined,
                      "text": undefined,
                    }
                  }
                  accessible={true}
                  centered={true}
                  collapsable={false}
                  focusable={true}
                  hitSlop={
                    {
                      "bottom": 6,
                      "left": 6,
                      "right": 6,
                      "top": 6,
                    }
                  }
                  onBlur={[Function]}
                  onClick={[Function]}
                  onFocus={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    [
                      {
                        "overflow": "hidden",
                      },
                      [
                        {
                          "alignItems": "center",
                          "flexGrow": 1,
                          "justifyContent": "center",
                        },
                        {
                          "borderRadius": 20,
                        },
                      ],
                    ]
                  }
                  testID="right-icon-adornment"
                >
                  <Text
                    accessibilityElementsHidden={true}
                    allowFontScaling={false}
                    importantForAccessibility="no-hide-descendants"
                    pointerEvents="none"
                    selectable={false}
                    style={
                      [
                        {
                          "color": "rgba(103, 80, 164, 1)",
                          "fontSize": 24,
                        },
                        [
                          {
                            "lineHeight": 24,
                            "transform": [
                              {
                                "scaleX": 1,
                              },
                            ],
                          },
                          {
                            "backgroundColor": "transparent",
                          },
                        ],
                        {
                          "fontFamily": "material-community",
                          "fontStyle": "normal",
                          "fontWeight": "normal",
                        },
                        {},
                      ]
                    }
                  >
                    ó°°
                  </Text>
                </View>
              </View>
            </View>
          </View>
        </View>
        <View
          style={
            {
              "height": 20,
            }
          }
        />
      </View>
      <View
        collapsable={false}
        style={
          {
            "backgroundColor": "rgba(98, 91, 113, 1)",
            "borderRadius": 20,
            "shadowColor": "#000",
            "shadowOffset": {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 0,
            "shadowRadius": 0,
          }
        }
        testID="button-container-outer-layer"
      >
        <View
          collapsable={false}
          style={
            {
              "backgroundColor": "rgba(98, 91, 113, 1)",
              "borderColor": "transparent",
              "borderRadius": 20,
              "borderStyle": "solid",
              "borderWidth": 0,
              "flex": undefined,
              "minWidth": 64,
              "shadowColor": "#000",
              "shadowOffset": {
                "height": 0,
                "width": 0,
              },
              "shadowOpacity": 0,
              "shadowRadius": 0,
            }
          }
          testID="button-container"
        >
          <View
            accessibilityRole="button"
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": false,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              [
                {
                  "overflow": "hidden",
                },
                {
                  "borderRadius": 20,
                },
              ]
            }
            testID="button"
          >
            <View
              style={
                [
                  {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "center",
                  },
                  undefined,
                ]
              }
            >
              <View
                style={
                  [
                    {
                      "marginLeft": 12,
                      "marginRight": -4,
                    },
                    {
                      "marginLeft": 16,
                      "marginRight": -16,
                    },
                    false,
                  ]
                }
                testID="button-icon-container"
              >
                <Text
                  accessibilityElementsHidden={true}
                  allowFontScaling={false}
                  importantForAccessibility="no-hide-descendants"
                  pointerEvents="none"
                  selectable={false}
                  style={
                    [
                      {
                        "color": "rgba(125, 82, 96, 1)",
                        "fontSize": 18,
                      },
                      [
                        {
                          "lineHeight": 18,
                          "transform": [
                            {
                              "scaleX": 1,
                            },
                          ],
                        },
                        {
                          "backgroundColor": "transparent",
                        },
                      ],
                      {
                        "fontFamily": "material-community",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      {},
                    ]
                  }
                >
                  ó°´¯
                </Text>
              </View>
              <Text
                numberOfLines={1}
                selectable={false}
                style={
                  [
                    {
                      "textAlign": "left",
                    },
                    {
                      "color": "rgba(28, 27, 31, 1)",
                      "writingDirection": "ltr",
                    },
                    [
                      {
                        "fontFamily": "System",
                        "fontSize": 14,
                        "fontWeight": "500",
                        "letterSpacing": 0.1,
                        "lineHeight": 20,
                      },
                      [
                        {
                          "marginHorizontal": 16,
                          "marginVertical": 9,
                          "textAlign": "center",
                        },
                        false,
                        {
                          "marginHorizontal": 24,
                          "marginVertical": 10,
                        },
                        undefined,
                        false,
                        {
                          "color": "rgba(125, 82, 96, 1)",
                          "fontFamily": "System",
                          "fontSize": 14,
                          "fontWeight": "500",
                          "letterSpacing": 0.1,
                          "lineHeight": 20,
                        },
                        undefined,
                      ],
                    ],
                  ]
                }
                testID="button-text"
              >
                APPLY PROMO CODE
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
  <Text
    style={
      [
        {
          "textAlign": "left",
        },
        {
          "color": "rgba(28, 27, 31, 1)",
          "writingDirection": "ltr",
        },
        [
          {
            "fontFamily": "System",
            "fontSize": 16,
            "fontWeight": "500",
            "letterSpacing": 0.15,
            "lineHeight": 24,
          },
          [
            {
              "color": "rgba(125, 82, 96, 1)",
            },
            {
              "marginTop": 20,
              "textAlign": "center",
            },
          ],
        ],
      ]
    }
  >
    Your basket is empty.
  </Text>
</View>
`;
